File: Medium/City With the Smallest Number of Neighbors at a Threshold Distance/city-with-the-smallest-number-of-neighbors-at-a-threshold-distance.py
#User function Template for python3

from typing import List
import heapq

class Solution:
    def findCity(self, n : int, m : int, edges : List[List[int]], distanceThreshold : int) -> int:
        # Create graph
        graph = [[] for _ in range(n)]
        for edge in edges:
            from_city, to_city, weight = edge
            graph[from_city].append((to_city, weight))
            graph[to_city].append((from_city, weight))  # Adding bidirectional edge
        
        min_reachable_cities = float('inf')
        optimal_city = -1

        # Iterate through each city
        for i in range(n):
            distances = [float('inf')] * n
            pq = [(0, i)]
            distances[i] = 0

            # Dijkstra's Algorithm
            while pq:
                curr_distance, curr_city = heapq.heappop(pq)

                for neighbor, neighbor_distance in graph[curr_city]:
                    new_distance = curr_distance + neighbor_distance
                    if new_distance <= distanceThreshold and new_distance < distances[neighbor]:
                        distances[neighbor] = new_distance
                        heapq.heappush(pq, (new_distance, neighbor))

            # Count reachable cities
            reachable_cities = sum(1 for d in distances if d <= distanceThreshold)

            # Update optimal city
            if reachable_cities < min_reachable_cities or (reachable_cities == min_reachable_cities and i > optimal_city):
                min_reachable_cities = reachable_cities
                optimal_city = i

        return optimal_city

#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
    T = int(input())
    for t in range(T):
        
        n, m = map(int, input().strip().split())
        edges = []
        for i in range(m):
            u, v, wt = map(int, input().strip().split())
            edges.append([u, v, wt])
        distanceThreshold = int(input())
        obj = Solution()
        ans = obj.findCity(n, m, edges, distanceThreshold)
        print(ans)
            

# } Driver Code Ends

File: Medium/City With the Smallest Number of Neighbors at a Threshold Distance/city-with-the-smallest-number-of-neighbors-at-a-threshold-distance.cpp
//{ Driver Code Starts
// Initial Template for C++

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// User function Template for C++
class Solution {
  public:
    int findCity(int n, int m, vector<vector<int>>& edges, int distanceThreshold) {
        vector<vector<pair<int, int>>> graph(n);
        for (const auto& edge : edges) {
            int from = edge[0];
            int to = edge[1];
            int weight = edge[2];
            graph[from].push_back({to, weight});
            graph[to].push_back({from, weight}); // Adding bidirectional edge
        }

        int minReachableCities = INT_MAX;
        int optimalCity = -1;

        // Iterate through each city
        for (int i = 0; i < n; ++i) {
            vector<int> distances(n, INT_MAX);
            priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
            pq.push({0, i});
            distances[i] = 0;

            // Dijkstra's Algorithm
            while (!pq.empty()) {
                int currDistance = pq.top().first;
                int currCity = pq.top().second;
                pq.pop();

                for (const auto& neighbor : graph[currCity]) {
                    int neighborCity = neighbor.first;
                    int neighborDistance = neighbor.second;
                    int newDistance = currDistance + neighborDistance;
                    if (newDistance <= distanceThreshold && newDistance < distances[neighborCity]) {
                        distances[neighborCity] = newDistance;
                        pq.push({newDistance, neighborCity});
                    }
                }
            }

            // Count reachable cities
            int reachableCities = 0;
            for (int j = 0; j < n; ++j) {
                if (distances[j] <= distanceThreshold) {
                    ++reachableCities;
                }
            }

            // Update optimal city
            if (reachableCities < minReachableCities || (reachableCities == minReachableCities && i > optimalCity)) {
                minReachableCities = reachableCities;
                optimalCity = i;
            }
        }

        return optimalCity;
    }
};


//{ Driver Code Starts.
int main() {
    int t;
    cin >> t;
    while (t--) {
        int n, m;
        cin >> n >> m;
        vector<vector<int>> adj;
        // n--;
        for (int i = 0; i < m; ++i) {
            vector<int> temp;
            for (int j = 0; j < 3; ++j) {
                int x;
                cin >> x;
                temp.push_back(x);
            }
            adj.push_back(temp);
        }

        int dist;
        cin >> dist;
        Solution obj;
        cout << obj.findCity(n, m, adj, dist) << "\n";
    }
}

// } Driver Code Ends